src/python/pants/bin/daemon_pants_runner.py

```diff
@@ -55,93 +69,63 @@ def exit(self, result=0, msg=None, *args, **kwargs):
       if msg:
-        NailgunProtocol.send_stderr(self._socket, msg)
+        NailgunProtocol.send_stderr(self._maybe_shutdown_socket.socket, msg)
 
       # Send an Exit chunk with the result.
-      NailgunProtocol.send_exit_with_code(self._socket, result)
+      # This will cause the client to disconnect from the socket.
+      NailgunProtocol.send_exit_with_code(self._maybe_shutdown_socket.socket, result)
 
       # Shutdown the connected socket.
-      teardown_socket(self._socket)
-    finally:
-      super(DaemonExiter, self).exit(result=result, *args, **kwargs)
-
+      teardown_socket(self._maybe_shutdown_socket.socket)
+      self._maybe_shutdown_socket.is_shutdown = True
 
-class _GracefulTerminationException(Exception):
-  """Allows for deferring the returning of the exit code of prefork work until post fork.
 
-  TODO: Once the fork boundary is removed in #7390, this class can be replaced by directly exiting
-  with the relevant exit code.
-  """
-
-  def __init__(self, exit_code=PANTS_FAILED_EXIT_CODE):
-    """
-    :param int exit_code: an optional exit code (defaults to PANTS_FAILED_EXIT_CODE)
-    """
-    super(_GracefulTerminationException, self).__init__('Terminated with {}'.format(exit_code))
-
-    if exit_code == PANTS_SUCCEEDED_EXIT_CODE:
-      raise ValueError(
-        "Cannot create _GracefulTerminationException with a successful exit code of {}"
-        .format(PANTS_SUCCEEDED_EXIT_CODE))
-
-    self._exit_code = exit_code
-
-  @property
-  def exit_code(self):
-    return self._exit_code
-
-
-class DaemonPantsRunner(ProcessManager):
+class DaemonPantsRunner(object):
   """A daemonizing PantsRunner that speaks the nailgun protocol to a remote client.
 
   N.B. this class is primarily used by the PailgunService in pantsd.
   """
 
   @classmethod
   def create(cls, sock, args, env, services, scheduler_service):
+    maybe_shutdown_socket = MaybeShutdownSocket(sock)
+
     try:
-      # N.B. This will temporarily redirect stdio in the daemon's pre-fork context
-      # to the nailgun session. We'll later do this a second time post-fork, because
-      # threads.
-      with cls.nailgunned_stdio(sock, env, handle_stdin=False):
+      # N.B. This will redirect stdio in the daemon's context to the nailgun session.
+      with cls.nailgunned_stdio(maybe_shutdown_socket, env, handle_stdin=False):
         options, _, options_bootstrapper = LocalPantsRunner.parse_options(args, env)
         subprocess_dir = options.for_global_scope().pants_subprocessdir
         graph_helper, target_roots, exit_code = scheduler_service.prefork(options, options_bootstrapper)
-        deferred_exc = None if exit_code == PANTS_SUCCEEDED_EXIT_CODE else _GracefulTerminationException(exit_code)
     except Exception:
-      deferred_exc = sys.exc_info()
       graph_helper = None
       target_roots = None
       options_bootstrapper = None
-      # N.B. This will be overridden with the correct value if options
```

Comment:
This comment should still be accurate I think? If it's not, then creating the `subprocess_dir` should no longer be necessary, and we can leave a TODO about removing it.